Adding compression to Hadoop save functions

pwendell opened the pull request:
In many HDFS clusters some or all of the data is stored with compression. This patch adds a hook to easily pass a compression codec class when calling saveAsTextFile, saveAsSequencefile, or saveAsHadoopFile, avoiding the need to create a custom JobConf for every save.

There are a few different ways one could factor this, since there is so much overloading and chaining in tall the saveAsXX. I tried to chose a way which didn't  dramatically change what is there now.

Fixes SPARK-763.

pwendell committed:
Adding compression to Hadoop save functions

pwendell committed:
Clean extra whitespace

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92/

ryanlecompte commented on core/src/main/scala/spark/PairRDDFunctions.scala:
```diff
@@ -583,11 +609,22 @@ class PairRDDFunctions[K: ClassManifest, V: ClassManifest](
     conf.setOutputValueClass(valueClass)
     // conf.setOutputFormat(outputFormatClass) // Doesn't work in Scala 2.9 due to what may be a generics bug
     conf.set("mapred.output.format.class", outputFormatClass.getName)
+    codec match {
```
Don't need to explicitly match on the option here, you can just:

``` scala
  codec.foreach { c =>
    conf.setCompressedMapOutput(true)
    conf.set("mapred.output.compress", true)
    conf.setMapOutputCompressorClass(c)
    conf.set("mapred.output.compression.codec", c.getCanonicalName)
    conf.set("mapred.output.compression.type", CompressionType.BLOCK.toString)
  }
```

mateiz replied:
Or another way is `for (c <- codec)`

pwendell replied:
Is this the canonical way to handle a binary option like this? I just don't like it beacuse it only saves 1 line of code, but it's much less explicit (i.e. it is using list semantics on something that really isn't a list).

rxin replied:
It is a bit weird, but it is the standard Scala way in which you treat the Option as a monad. I don't like this either, but it is sort of the community "standard".

ryanlecompte commented:
Nice!

mateiz commented:
Patrick, this is great, but can you also add it to the Java API?

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94/

mridulm commented:
This is pretty neat, thanks !
This helps us quite a lot :-)

pwendell committed:
Monads FTW

pwendell committed:
Adding Java-API version of compression codec

pwendell commented:
I added a Java-API version. In general this is less of a "win" for the Java-API, because you already have to pass several arguments to the Hadoop saveAsX functions.

I also cleaned up up the Option thing to use Scala style.

pwendell commented:
Jenkins, test this please

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95/

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97/

pwendell committed:
Removing print statement

mateiz commented:
Thanks Patrick; this looks great.

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98/

pwendell commented:
I had to abort this test because it was running for 3 days.
